lib/modules.nix

```diff
@@ -807,6 +807,9 @@ let
 
     in warnDeprecation opt //
       { value = builtins.addErrorContext "while evaluating the option `${showOption loc}':" value;
+        # The merged value before applying the options `apply` function to it.
+        # In general though, `apply` should not be used, it's an anti-pattern.
+        beforeApply = res.mergedValue;
```

Comment:
Another angle is the current impossibility of using https://github.com/NixOS/nixpkgs/pull/257511 to simplify `submodule`. This can't currently be done, in large part because submodules have *two* values:
- the value also known as `config`
- the value with `_type = "configuration"`, aka the `evalModules` result, _containing `config`_, as well as `options` and other "meta" or "out of band" things.

What if `opt.beforeApply` was already thing, `submodule` returned the `_type = "configuration"`, and by convention, each submodule-typed option had `apply = x: x.config` or `apply = x: removeAttrs ["_module"] x.config`?
While that would be a little cumbersome to write, especially for more complicated hierarchies with multiple submodule/attrsOf nestings, it would be more capable, more efficient (fewer `removeAttrs` calls that copy almost everything), and possible to orthogonalize `submodule` into individually usable types, such as more or less attempted in https://github.com/NixOS/nixpkgs/pull/257511.

Reply:
Sounds interesting, though I don't think having `beforeApply` would be blocking that.

Once we deprecate `apply` entirely, `beforeApply` will vanish too :)

Reply:
> Once we deprecate `apply` entirely, `beforeApply` will vanish too :)

I think `apply` will stick around until we replace modules themselves. Do we really want to inflict its removal on users?